pkg/distributor/distributor.go

```diff
@@ -991,6 +992,64 @@ func (d *Distributor) MetricsForLabelMatchers(ctx context.Context, from, through
 	return result, nil
 }
 
+func (d *Distributor) MetricsForLabelMatchersStream(ctx context.Context, from, through model.Time, matchers ...*labels.Matcher) ([]metric.Metric, error) {
```

Comment:
There is quite a bit of duplicate code to `MetricsForLabelMatchers`, I wonder if we can refactor out some common code.